### PR TITLE
Make max error in approx_set tunable

### PR DESCRIPTION
**Pull request by blrnw3**
Adds a new version: approx_set(x, e)
Also for empty_approx_set(e)

This brings it in line with the approx_distinct functionality, although I kept the existing difference in default max standard error (0.01625 vs 0.023 i.e. 4096 vs 2048 buckets).

Issue: #12506

### CODE REVIEW

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -7230,6 +7276,22 @@ public void testMergeEmptyNonEmptyApproxSet()
...
+    public void testMergeEmptyNonEmptyApproxSetWithSameMaxError()
+    {
+        MaterializedResult actual = computeActual("SELECT cardinality(merge(c)) FROM (SELECT create_hll(custkey, 0.1) c FROM orders UNION ALL SELECT empty_approx_set(0.1))");
+        MaterializedResult expected = resultBuilder(getSession(), BIGINT)
+                .row(1046L)
+                .build();
+        assertEquals(actual.getMaterializedRows(), expected.getMaterializedRows());
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class, expectedExceptionsMessageRegExp = "Cannot merge HLLs with different number of buckets.*")
```
Comment (highker) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: `RuntimeException`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateCountDistinctAggregation.java [hunk 2]
```diff
@@ -127,7 +126,7 @@ public static void input(BooleanDistinctState state, @SqlType(StandardTypes.BOOL
         state.setByte((byte) (state.getByte() | (value ? 1 : 2)));
     }
 
-    private static HyperLogLog getOrCreateHyperLogLog(HyperLogLogState state, double maxStandardError)
+    static HyperLogLog getOrCreateHyperLogLog(HyperLogLogState state, double maxStandardError)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateCountDistinctAggregation.java]: Maybe extract the helper into a common utils rather than keep it in one particular aggregation function. For example, `HyperLogLogUtils`, `HyperLogLogAggregations`, ...

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 3]
```diff
@@ -4048,7 +4048,7 @@ public void testWindowNoChannels()
     public void testInvalidWindowFunction()
     {
         assertQueryFails("SELECT abs(x) OVER ()\n" +
-                "FROM (VALUES (1), (2), (3)) t(x)",
+                        "FROM (VALUES (1), (2), (3)) t(x)",
```
Comment (highker) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: unrelated change

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateSetAggregation.java [hunk 4]
```diff
@@ -31,53 +31,53 @@
 @AggregationFunction("approx_set")
 public final class ApproximateSetAggregation
 {
-    private static final int NUMBER_OF_BUCKETS = 4096;
     private static final AccumulatorStateSerializer<HyperLogLogState> SERIALIZER = StateCompiler.generateStateSerializer(HyperLogLogState.class);
+    public static final double DEFAULT_STANDARD_ERROR = 0.01625;
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateSetAggregation.java]: swap the order of this variable with the previous one.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/HyperLogLogFunctions.java [hunk 5]
```diff
@@ -38,6 +39,14 @@ public static long cardinality(@SqlType(StandardTypes.HYPER_LOG_LOG) Slice seria
...
     public static Slice emptyApproxSet()
     {
-        return ApproximateSetAggregation.newHyperLogLog().serialize();
+        return HyperLogLog.newInstance(ApproximateCountDistinctAggregation.standardErrorToBuckets(ApproximateSetAggregation.DEFAULT_STANDARD_ERROR)).serialize();
+    }
+
+    @ScalarFunction
+    @Description("an empty HyperLogLog instance with the specified max standard error")
+    @SqlType(StandardTypes.HYPER_LOG_LOG)
+    public static Slice emptyApproxSet(@SqlType(StandardTypes.DOUBLE) double maxStandardError)
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/HyperLogLogFunctions.java]: Maybe an overkill: the empty HLL may have "unknown" bucket count. The bucket count remains unknown when merging with an unknown-bucket HLL or coalesced with the other one's bucket count. But given the complication, I don't have strong opinion having this.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateSetAggregation.java [hunk 6]
```diff
@@ -31,53 +31,53 @@
...
-        if (hll == null) {
-            hll = newHyperLogLog();
-            state.setHyperLogLog(hll);
-            state.addMemoryUsage(hll.estimatedInMemorySize());
-        }
-        return hll;
+        HyperLogLog hll = ApproximateCountDistinctAggregation.getOrCreateHyperLogLog(state, maxStandardError);
+        state.addMemoryUsage(-hll.estimatedInMemorySize());
+        hll.add(value);
+        state.addMemoryUsage(hll.estimatedInMemorySize());
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ApproximateSetAggregation.java]: It will throw an `IllegalArgumentException` in the combine method below, which will show up as an internal error in Presto (something that shouldn't happen), where this should show up as a user error.  I think it should be repackaged: `throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e.getMessage(), e);`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/HyperLogLogFunctions.java (hunk 5 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/HyperLogLogFunctions.java]: I agree with it, this would be a nice thing to do, but if it's too complicated we can skip it

File: presto-tests/src/main/java/com/facebook/presto/tests/CreateHll.java [hunk 7]
```diff
@@ -27,7 +29,16 @@ private CreateHll() {}
     @SqlType(StandardTypes.HYPER_LOG_LOG)
     public static Slice createHll(@SqlType(StandardTypes.BIGINT) long value)
     {
-        HyperLogLog hll = HyperLogLog.newInstance(4096);
+        HyperLogLog hll = HyperLogLog.newInstance(ApproximateCountDistinctAggregation.standardErrorToBuckets(ApproximateSetAggregation.DEFAULT_STANDARD_ERROR));
```
Comment (tdcmeehan) [inline presto-tests/src/main/java/com/facebook/presto/tests/CreateHll.java]: I would add some static imports: `HyperLogLog hll = HyperLogLog.newInstance(standardErrorToBuckets(DEFAULT_STANDARD_ERROR));`

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 8]
```diff
@@ -7230,6 +7276,22 @@ public void testMergeEmptyNonEmptyApproxSet()
...
+                .build();
+        assertEquals(actual.getMaterializedRows(), expected.getMaterializedRows());
+    }
+
+    @Test(expectedExceptions = IllegalArgumentException.class, expectedExceptionsMessageRegExp = "Cannot merge HLLs with different number of buckets.*")
+    public void testMergeEmptyNonEmptyApproxSetWithDifferentMaxError()
+    {
+        computeActual("SELECT cardinality(merge(c)) FROM (SELECT create_hll(custkey, 0.1) c FROM orders UNION ALL SELECT empty_approx_set(0.2))");
+    }
+
```
Comment (tdcmeehan) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Can you add a test for error outside the bounds, like 1?  Also if it throws `IllegalArgumentException`, can you repackage it so it is a Presto exception?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/HyperLogLogFunctions.java (hunk 5 above)
Comment (blrnw3) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/HyperLogLogFunctions.java]: Yes that would be nice, but we don't currently have access to the bucket count of an HLL so I don't think it is easy yet. I opened a PR to allow this so in future this could be possible: https://github.com/airlift/airlift/pull/724

**Comment by blrnw3** [comment]
Thanks both. Comments have been addressed.

**Comment by blrnw3** [comment]
@tdcmeehan please can I get a review on my changes?

**Comment by blrnw3** [comment]
Thanks!
